text/0000-repr-packed-unsafe-ref.md

```diff
@@ -0,0 +1,438 @@
+}
+```
+
+On playpen, it prints:
+
+```
+before
+middle
+playpen: application terminated abnormally with signal 4 (Illegal instruction)
+```
+
+That is, the `bad` variable is causing the CPU to fault. The `let`
+statement is (in pseudo-Rust) behaving like `let bad =
+load_with_alignment(&val.y.0, align_of::<f32x4>());`, but the
+alignment isn't satisfied. (The `ok` line is compiled to a `movupd`
+instruction, while the `bad` is compiled to a `movapd`: `u` ==
+unaligned, `a` == aligned.)
+
+(NB. The use of SIMD types in the example is just to be able to
+demonstrate the problem on x86. That platform is generally fairly
+relaxed about pointer alignments and so SIMD & its specialised `mov`
+instructions are the easiest way to demonstrate the violated
+assumptions at runtime. Other platforms may fault on other types.)
+
+Being able to assume that accesses are aligned is useful, for
+performance, and almost all references will be correctly aligned
+anyway (`repr(packed)` types and internal references into them are
+quite rare).
+
+The problems with unaligned accesses can be avoided by ensuring that
+the accesses are actually aligned (e.g. via runtime checks, or other
+external constraints the compiler cannot understand directly). For
+example, consider the following
+
+```rust
+#[repr(packed, C)]
+struct Bar {
+    x: u8,
+    y: u16,
+    z: u8,
+    w: u32,
+}
+```
+
+Taking a reference to some of those fields may cause undefined
+behaviour, but not always. It is always correct to take
+a reference to `x` or `z` since `u8` has alignment 1. If the struct
+value itself is 4-byte aligned (which is not guaranteed), `w` will
+also be 4-byte aligned since the `u8, u16, u8` take up 4 bytes, hence
+it is correct to take a reference to `w` in this case (and only that
+case). Similarly, it is only correct to take a reference to `y` if the
+struct is at an odd address, so that the `u16` starts at an even one
+(i.e. is 2-byte aligned).
+
+# Detailed design
+
+It is `unsafe` to take a reference to the field of a `repr(packed)`
+struct. It is still possible, but it is up to the programmer to ensure
+that the alignment requirements are satisfied. Referencing
+(by-reference, or by-value) a subfield of a struct (including indexing
+elements of a fixed-length array) stored inside a `repr(packed)`
+struct counts as taking a reference to the `packed` field and hence is
+unsafe.
+
+It is still legal to manipulate the fields of a `packed` struct by
+value, e.g. the following is correct (and not `unsafe`), no matter the
+alignment of `bar`:
+
+```rust
+let bar: Bar = ...;
+
+let x = bar.y;
+bar.w = 10;
+```
+
+It is illegal to store a type `T` implementing `Drop` (including a
+generic type) in a `repr(packed)` type, since the destructor of `T` is
+passed a reference to that `T`. The crater run (see appendix) found no
+crate that needs to use `repr(packed)` to store a `Drop` type (or a
+generic type). The generic type rule is conservatively approximated by
```

Comment:
Haha, rustc isn't covered by crater.

It looks like that case could be handled by a trait that uses a monomorphic `#[repr(packed)]` internally to do an unaligned load, since it is just used to load `{u,i}{8,16,32,64,usize}` data?